# Исследование надежности заемщиков


**Цель исследования:** Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок.


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
#Импортирование библеотеки и чтение файла с данными
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
# получение первых 20 строк таблицы head
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
# получение общей информации о данных в таблице data
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


**Вывод:**

В каждой строке таблицы — данные клиенте банка. Часть колонок описывает клиента: дети, общий трудовой стаж, возвраст клиента, уровень образования,семейное положение, пол клиента, тип занятости, ежемесячных доход. Часть данных помогают для классификации данных:идентификатор уровня образования,идентификатор семейного положения. Последняя часть данных рассказывают о кредите клиента:имел ли задолженность по возврату кредитов,цель получения кредита.

Предварительно можно утверждать, что данных достаточно для проверки цели исследования. Но встречаются пропуски в данных, также несоотвествие типов данных и нехватки категоризации данных.

Чтобы двигаться дальше, нужно устранить проблемы в данных.

## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
# подсчёт пропусков
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
# заполнение пропусков медианными значениям в столбце total_income
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
# Замена отрицательных данных на положительные в столбце days_employed
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
# Медианное значение трудового стажа для каждого типа занятости
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
# просмотр уникальных значений столбца children
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
# Удаление аномальных значений в стоблце children
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
# проверка на отсутствие аномальных значений
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
# заполнение пропусков медианными значениям в столбце days_employed
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
# проверка на отсутствие пропусков
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
# Изменение типа данных в столбце total_income
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
# обработка неявных дубликатов, приведение данных в нижний регистр в стоблце education
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
# подсчет дубликатов
data.duplicated().sum()

71

In [16]:
# удаление дубликатов
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
# создание фукнции для категоризации клиентов по доходности
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
#применение функции к таблице 
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
# вывод уникальных значений в столбце purpose
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
# создание функции для категоризация данных по столбцу purpose
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
#применение функции к таблице 
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
# создание функции для группировки данных

def rate(row,parametr):
    data = row.groupby(parametr).agg({'debt':'sum'})
    data['count'] = row.groupby(parametr)['debt'].count()
    data['rate'] = data['debt'] / data['count'] * 100
    return data

In [24]:
#применение функции и создание таблицы data_children_dependence
data_children_dependence = rate(data,'children')
data_children_dependence.sort_values(by='rate')

debt  count      rate
children                       
5            0      9  0.000000
0         1063  14091  7.543822
3           27    330  8.181818
1          444   4808  9.234609
2          194   2052  9.454191
4            4     41  9.756098

**Вывод:** Семьи с количеством детей  5,4 и 3 анализировать доставляет трудности, так как мало данных для их анализа. Более того семьи без детей берут кредит намного больше, также задолженность составляет наименьший процент. Чем больше детей появляется в семье, тем кредитов берется меньше и процент задолженности растет. Тем самым разница количество кредитов между 0 и 2 детьми в семье составляет 12 039, а разница в задолженности 1,91 процент.

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

<a id="title2"></a>

In [25]:
# применение функции и создание таблицы data_family_dependence
data_family_dependence = rate(data,'family_status')
data_family_dependence.sort_values(by='rate')

debt  count      rate
family_status                               
вдовец / вдова           63    951  6.624606
в разводе                84   1189  7.064760
женат / замужем         927  12261  7.560558
гражданский брак        385   4134  9.313014
Не женат / не замужем   273   2796  9.763948

**Вывод:** В данной выборке все группы имеют достаточно данных для анализа. Группа вдовец/вдова,в разводе,женат/замужем выплачивают кредиты исправно, нежели группы, которые находятся в гражданском браке или без отношений. Возможно отношения или прошлые отношения влияют на дисциплированность людей.

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

<a id="title3"></a>

In [26]:
# применение функции и создание таблицы data_income_dependence
data_income_dependence = rate(data,'total_income_category')
data_income_dependence.sort_values(by = 'rate')

debt  count      rate
total_income_category                       
D                        21    349  6.017192
B                       354   5014  7.060231
A                         2     25  8.000000
C                      1353  15921  8.498210
E                         2     22  9.090909

**Вывод:** В выборке группа D,А,Е имеют мало данных для анализа. Чем меньше у тебя дохода, тем больше кредитов ты берешь, также это влияет на процент задолженности. Это видно при сравнении группы B и С, где группа В имея больший доход, берет на 10907 меньше кредитов и платит исправнее на 1,43 процента.

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

<a id="title4"></a>

In [27]:
# применение функции и создание таблицы data_purpose
data_purpose = rate(data,'purpose_category')
data_purpose.sort_values(by = 'rate')

debt  count      rate
purpose_category                               
операции с недвижимостью   780  10751  7.255139
проведение свадьбы         183   2313  7.911803
получение образования      369   3988  9.252758
операции с автомобилем     400   4279  9.347978

**Вывод:** 50% кредитов приходится на операции с недвижимостью, также такие кредиты выплачиваются исправно. Самые ненадежные кредиты являются операции с автомобилем. 

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причина пропусков может быть как и человеческим фактором, так и технологической. Так как количество пропусков одинакова(2174), сотрудник мог их не заполнить или же где то система не собирает какие то данные. Более того, возможно при переносе данных из таблиц, сбой системы, обновления могли повлиять на пропуски.


#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Так как удаление  пропусков может повлиять на общий анализ данных, нельзя их просто удалить, так как можно подставить данные, ибо мы имеем примерные схожие данные. Более того, мы не можем оставить пропуски, так как это повлияет на дальнейший анализ, возможны ошибки. Среднеарифметическое может сильно подвергаться к изменениям из за экстремальных максимальных и минимальных значений. Тем самым медианные значение не будет сильно зависеть от экстремальных данных. Таким образом, данный метод является оптимальным, так как помогает заполнить данные схожими данными по категориям и не дает большую разницу с вероятно реальными данными. 

### Шаг 4: общий вывод.

Были исследованы данные о клиентах банка и их задолженности. 

**Что было сделано?**
Были заполнены пустые данные медианными значениями каждой группы в столбцах total_income(доход) и days_employed(стаж работы). Также в столбце days_employed(стаж работы) отрицательные значения были заменены на положительные, так как скорее всего это ошибка перевода, прочтение файла или человеческий фактор. Были удалены явные и неявные дубликаты в таблице. Также были созданы два новых столбца для категоризации данных: 1 - категоризация доходов клиентов, 2 - категоризация по целям взятие кредита. 

**Выявленные зависимости**
1.Чем больше детей, тем больше кредитов семьи берут кредиты. Также исправность платежа по кредиту также снижается.
Подробнее ознакомится с таблицей и выводом можете [тут](#title1).

2.Группа вдовец/вдова имеет меньше все задержек по платежу кредита, процент задержки составляет 6.62%. Группа которая имеет больше всего процентых задержек это Не женат / не замужем их процент составляет 9.76%. На группу женат / замужем приходится наибольшее количество кредитов - 12261
Подробнее ознакомится с таблицей и выводом можете [тут](#title2)

3.Чем меньше у тебя дохода, тем больше кредитов ты берешь, также это влияет на процент задолженности.Так у группы С количество кредитов 15 921, а процент неисправного платежа 8,49, а группы В количество кредитов 5014, а процент неисправного платежа составляет 7,04. Однако, это не окончательный вывод, стоит больше получить данные, так как этот вывод строился из анализа только двух групп.
Подробнее ознакомится с таблицей и выводом можете [тут](#title3)

4.Меньше всего просрочек имеет группа с целью на недвижимость(7.25%), больше всего просрочек имеет группа с целью на автомобиль(9.34%). Наибольшая выборка данных имеет группа с целью на недвижимость - 10751. 
Подробнее ознакомится с таблицей и выводом можете [тут](#title4)

**Портрет надежного заемщика**
Это заемщик с доходом уровня B, также данный заемщик не имеет детей, находился или находится в браке. Также цель его кредита желательно должна иметь цель недвижимость или для проведение свадьбы.

**Портрет рискового заемщика**
Это заемщик с доходом уровня С, также данный заемщик имеет более 1 ребенка, не находится в отношениях или в гражданском браке. Также их цель кредита является операции с автомобилем или образования. 

**Рекомендации**
Стоит больше собрать данных по другим группах по уровню дохода клиентов. Так как данный вывод может быть ошибочным, ибо были проанализированы только две группы из 5. 
